### ISSUE TITLE
Logspout integration for logging through Docker

### ISSUE THREAD

**Issue by sandys** [grafana/loki#82]
A lot (most?) deployments of applications happen through docker. These use frameworks like Docker Swarm or Kubernetes to deploy. The deployment has little awareness of the underlying hardware/OS.

Any logging solution must integrate into Docker deployment frameworks.

A popular/generic solution is Logspout - https://github.com/gliderlabs/logspout

It is now adopted by a lot of log providers - https://github.com/logdna/logspout and https://www.loggly.com/docs/docker-logging-logspout/

**Comment by Kuqd** [comment]
Does the [docker driver](https://github.com/grafana/loki/blob/master/cmd/docker-driver/README.md) is enough  or does logsprout has some other interesting features ?

**Comment by megamorf** [comment]
@Kuqd Its main advantage is that it keeps the regular docker log output intact whereas specifying a docker logging driver per container or host breaks `docker logs` entirely. It can also use multiple log "adapters" in parallel, e.g. send logs to graylog and syslog. In addition, it also offers routes to direct certain events to a different log system based on container labels.

Unfortunately, gliderlabs have significantly neglected its user base (i.e. lots of uncommented issues and pending pull requests). However, it'd still be a great addition to have since this would allow us to test loki while still sending all of our logs to our present log collector.

**Comment by slim-bean** [comment]
This issue has a lot of thumbs up, is this something people would still like to see?

**Comment by trevorwhitney** [comment]
Does @jeschkies [work](https://github.com/grafana/loki/pull/4911) on adding a docker target to promtail solve this problem?

**Comment by megamorf** [comment]
@trevorwhitney Nope, at least for now the promtail docker target implementation is incomplete. Docker will split up long events logged by the container application into multiple lines but promtail fails to reassemble those. It essentially results in broken JSON objects arriving in Loki.

See https://github.com/grafana/loki/issues/2281

**Comment by megamorf** [comment]
> This issue has a lot of thumbs up, is this something people would still like to see?

And yes, this is something that I'm still interested to see. Right now I have logspout sending container events to Graylog and would like to send those to Loki in parallel. That would enable me to prepare for shifting everything to loki seamlessly.

**Comment by trevorwhitney** [comment]
@megamorf I believe @jeschkies' implementation correctly adds those labels, but I haven't looked closely enough at it to know for sure.

**Comment by jeschkies** [comment]
@megamorf, yes the container labels are added. I'm not sure what you mean with host labels. WE use the service discovery from https://prometheus.io/docs/prometheus/latest/configuration/configuration/#docker_sd_config.

**Comment by megamorf** [comment]
> @megamorf, yes the container labels are added. I'm not sure what you mean with host labels. WE use the service discovery from https://prometheus.io/docs/prometheus/latest/configuration/configuration/#docker_sd_config.

You can also add common labels to the Docker daemon.json, see: 
- https://docs.docker.com/engine/reference/commandline/dockerd/#daemon-configuration-file
- https://docs.docker.com/config/containers/logging/configure/

A typical use case is to assign certain Docker hosts a label for a cluster, region or environment.

**Comment by cstyan** [comment]
@megamorf We're currently reevaluating promtails position as a project within Grafana Labs. Internally we're actually using the [Agent](https://github.com/grafana/agent) for both metrics and logs collection at this point. While we haven't made a formal decision yet, we expect in the near future that all new feature work will be done in the Agent's log collection pipelines rather than in Promtail. Would you be interested in opening an issue there for this feature?

**Comment by bboreham** [comment]
Closing as promtail is no longer part of Loki; we have a supported solution for [Docker](https://grafana.com/docs/alloy/latest/monitor/monitor-docker-containers/#configure-logging) and [Kubernetes](https://grafana.com/docs/alloy/latest/collect/logs-in-kubernetes/) in Grafana Alloy and there is now a cross-vendor standard to send telemetry - [OTLP](https://opentelemetry.io/docs/concepts/signals/logs/).
